sycl/test/basic_tests/parallel_for_range.cpp

```diff
@@ -1,93 +1,616 @@
+          nd_range<3>(range<3>(16, 16, 16), range<3>(8, 8, 8)),
+          [=](nd_item<3>) { reqd_wg_size_helper(); });
     });
-    q.wait_and_throw();
-    assert(false && "Should have thrown exception");
-  } catch (nd_range_error e) {
-    // We expect an error to be thrown!
+    Q.wait_and_throw();
+    std::cout << "Test case ReqdWGSizeNegativeA failed: no exception has been "
+                 "thrown\n";
+    return 1; // We shouldn't be here, exception is expected
+  } catch (nd_range_error &E) {
+    if (string_class(E.what()).find(
+            "Specified local size doesn't match the required work-group size "
+            "specified in the program source") == string_class::npos) {
+      std::cout
+          << "Test case ReqdWGSizeNegativeA failed: unexpected exception: "
+          << E.what() << std::endl;
+      return 1;
+    }
+  } catch (runtime_error &E) {
+    std::cout << "Test case ReqdWGSizeNegativeA failed: unexpected exception: "
+              << E.what() << std::endl;
+    return 1;
+  } catch (...) {
+    std::cout << "Test case ReqdWGSizeNegativeA failed: something unexpected "
+                 "has been caught"
+              << std::endl;
+    return 1;
   }
 
-  // parallel_for, 100 global, 4 local -> pass.
+  string_class OCLVersionStr = D.get_info<info::device::version>();
+  assert(OCLVersionStr.size() >= 10 &&
+         "Unexpected device version string"); // strlen("OpenCL X.Y")
+  const char *OCLVersion = &OCLVersionStr[7]; // strlen("OpenCL ")
+  if (OCLVersion[0] == '1' || (OCLVersion[0] == '2' && OCLVersion[2] == '0')) {
+    // parallel_for, (16, 16, 16) global, null local, reqd_wg_size(4, 4, 4) //
+    // -> fail
+    try {
+      Q.submit([&](handler &CGH) {
+        CGH.parallel_for<class ReqdWGSizeNegativeB>(
+            range<3>(16, 16, 16), [=](item<3>) { reqd_wg_size_helper(); });
+      });
+      Q.wait_and_throw();
+      std::cout
+          << "Test case ReqdWGSizeNegativeB failed: no exception has been "
+             "thrown\n";
+      return 1; // We shouldn't be here, exception is expected
+    } catch (nd_range_error &E) {
+      if (string_class(E.what()).find(
+              "OpenCL 1.x and 2.0 requires to pass local size argument even if "
+              "required work-group size was specified in the program source") ==
+          string_class::npos) {
+        std::cout
+            << "Test case ReqdWGSizeNegativeB failed: unexpected exception: "
+            << E.what() << std::endl;
+        return 1;
+      }
+    } catch (runtime_error &E) {
+      std::cout
+          << "Test case ReqdWGSizeNegativeB failed: unexpected exception: "
+          << E.what() << std::endl;
+      return 1;
+    } catch (...) {
+      std::cout << "Test case ReqdWGSizeNegativeB failed: something unexpected "
+                   "has been caught"
+                << std::endl;
+      return 1;
+    }
+  }
+
+  // Positive test-cases that should pass on any underlying OpenCL runtime
```

Comment:
Don't see much value in "positive" test cases here. I believe they are already covered in other tests.

Reply:
I've removed a few of them - the rest seems useful to me (related to reqd_work_group_size attribute and custom build options)